Compile error when using KDBindings in combination with Qt

Hi!
I wanted to use KSBindings in a project where I also use Qt. At a certain point I needed to include kdbindings/signal.h as well as QObject. This then results in a compile error (on gcc as well as clang). As I investigated further it turned out that the order of includes matter. If KDBindings is included first, it works. If QObject is first then it doesn’t.
I was surprised and wanted to ask if this is a known limitation.

The compile error looks like the following:

```
[  9%] Building CXX object CMakeFiles/workshop_episode_001.dir/src/Translator.cpp.o
In file included from src/Translator.cpp:3:
In file included from include/IResourceManager.hpp:3:
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:285:23: error: expected ')'
        void emit(Args... p) const
                      ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:285:18: note: to match this '('
        void emit(Args... p) const
                 ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:285:19: error: declaration of 'Args' shadows template parameter
        void emit(Args... p) const
                  ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:213:22: note: template parameter is declared here
template<typename... Args>
                     ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:285:19: error: field has incomplete type 'void'
        void emit(Args... p) const
                  ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:285:29: error: expected ';' at end of declaration list
        void emit(Args... p) const
                            ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:308:49: error: use of undeclared identifier 'Connection'
        mutable Private::GenerationalIndexArray<Connection> m_connections;
                                                ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:488:19: error: expected ')'
    void emit(Args... p) const
                  ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:488:14: note: to match this '('
    void emit(Args... p) const
             ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:488:15: error: declaration of 'Args' shadows template parameter
    void emit(Args... p) const
              ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:213:22: note: template parameter is declared here
template<typename... Args>
                     ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:488:15: error: field has incomplete type 'void'
    void emit(Args... p) const
              ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:488:25: error: expected ';' at end of declaration list
    void emit(Args... p) const
                        ^
conan_cache/.conan/data/kdbindings/1.0.1/conan_toolchain_catalog/build/package/e7a4df623f2f69d6974a12334b94709d4f79466f/include/kdbindings/signal.h:259:13: error: unknown type name 'Connection'
            Connection *connection = m_connections.get(id);
            ^
10 errors generated.
make[3]: *** [CMakeFiles/workshop_episode_001.dir/build.make:161: CMakeFiles/workshop_episode_001.dir/src/Translator.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:1247: CMakeFiles/workshop_episode_001.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1254: CMakeFiles/workshop_episode_001.dir/rule] Error 2
make: *** [Makefile:712: workshop_episode_001] Error 2 
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,6 +41,7 @@ option(${PROJECT_NAME}_TESTS "Build the tests" ON)
 option(${PROJECT_NAME}_EXAMPLES "Build the examples" ON)
 option(${PROJECT_NAME}_DOCS "Build the API documentation" OFF)
 option(${PROJECT_NAME}_ERROR_ON_WARNING "Enable all compiler warnings and treat them as errors" OFF)
+option(${PROJECT_NAME}_QT_NO_EMIT "Qt Compatibility: Disable Qt's `emit` keyword" OFF)
 
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/ECM/modules)

--- a/README.md
+++ b/README.md
@@ -21,6 +21,16 @@ Find more information at:
  * [detailed browsable API reference](https://docs.kdab.com/kdbindings/latest/getting-started/)
  * [our assorted example programs](https://github.com/KDAB/KDBindings/blob/main/examples)
 
+Compatibility with Qt
+=====================
+Because Qt defines `emit` as a keyword, Qt is by default incompatible with KDBindings, which uses `emit` as a function name.
+
+This can only be fixed by defining `QT_NO_EMIT` or `QT_NO_KEYWORDS` and simply omitting `emit` in your Qt code.
+
+To make this easy, KDBindings includes a CMake option: `KDBindings_QT_NO_EMIT`, which defines this for all dependents of KDBindings.
+
+See also: [#79](https://github.com/KDAB/KDBindings/issues/79)
+
 Contact
 =======
 * Visit us on GitHub: https://github.com/KDAB/KDBindings

--- a/src/kdbindings/CMakeLists.txt
+++ b/src/kdbindings/CMakeLists.txt
@@ -40,6 +40,9 @@ set_target_properties(KDBindings PROPERTIES
 if(KDBINDINGS_ENABLE_WARN_UNUSED)
   target_compile_definitions(KDBindings INTERFACE KDBINDINGS_ENABLE_WARN_UNUSED=1)
 endif()
+if(KDBindings_QT_NO_EMIT)
+  target_compile_definitions(KDBindings INTERFACE QT_NO_EMIT)
+endif()
 
 target_include_directories(KDBindings INTERFACE
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/..>

--- a/src/kdbindings/signal.h
+++ b/src/kdbindings/signal.h
@@ -17,6 +17,16 @@
 #include <type_traits>
 #include <utility>
 
+#ifdef emit
+static_assert(false, "KDBindings is not compatible with Qt's 'emit' keyword.\n"
+                     "To use KDBindings with Qt, please define QT_NO_EMIT to disable Qt's 'emit' keyword.\n"
+                     "If you're using CMake you can set KDBindings_QT_NO_EMIT to ON to add this.");
+
+// Undefine emit to suppress more compiler errors after the static_assert failure.
+// Otherwise the compiler would drown our custom error message with more errors.
+#undef emit
+#endif
+
 #include <kdbindings/connection_evaluator.h>
 #include <kdbindings/genindex_array.h>
 #include <kdbindings/utils.h>
@@ -29,7 +39,6 @@
  * All public parts of KDBindings are members of this namespace.
  */
 namespace KDBindings {
-
 /**
  * @brief A Signal provides a mechanism for communication between objects.
  *

